### PR TITLE
Add a bucket to the provider for deployments.

### PR DESCRIPTION
**Pull request by DavidTanner**
## What did you implement:

***Implementing Issue:*** #1513 

## How did you implement it:

Added a check for `deploymentBucket ` on the provider object.  If it is there, then remove the S3 resource from the CloudFormation setup.  Also all deploys are now prefixed with `serverless/${serviceName}/` so multiple deploys can happen without deleting other services.

## How can we verify it:
Use a config similar to this: 
```json
service: pr1513
provider:
  name: aws
  runtime: nodejs4.3
  stage: test
  region: us-west-2
  deploymentBucket: com.serverless.deploys
```
After the deploy you will see the S3 resource created initially, then removed from the CF file and the bucket created and deleted, finally in `com.serverless.deploys` there will be a new folder `serverless/pr1513/${dateString}/` that contains the deployment files.

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [x] Leave a comment that this is ready for review once you've finished the implementation

Also prefix deployments so they can co-exist with other files

### CODE REVIEW

**Comment by DavidTanner** [comment]
This should be ready to merge.

**Comment by eahefnawy** [comment]
@DavidTanner does this also support **creating** a bucket based on a custom name? or the bucket must exist?

**Comment by DavidTanner** [comment]
@eahefnawy If you want to override the current process of creating a new bucket, then just add the relevant section to the serverless.yml file.  This PR expects the bucket to already exist, and be in the same region as the Lambda function.

File: lib/plugins/aws/deploy/compile/functions/index.js [hunk 1]
```diff
@@ -112,6 +112,17 @@ class AwsCompileFunctions {
         throw new Error(`No artifact path is set for function: ${functionName}`);
       }
 
+      if (this.serverless.service.provider.deploymentBucket) {
```
Comment (eahefnawy) [inline lib/plugins/aws/deploy/compile/functions/index.js]: could you please add some validation logic for the bucket name along with some friendly error messages? 😊

File: lib/plugins/aws/deploy/compile/functions/index.js [hunk 2]
```diff
@@ -112,6 +112,17 @@ class AwsCompileFunctions {
...
+        const deploymentBucket = this.serverless.service.provider.deploymentBucket;
```
Comment (eahefnawy) [inline lib/plugins/aws/deploy/compile/functions/index.js]: also we should use the `getServerlessDeploymentBucketName()` you updated here imo to make use of the region validation logic.

**Comment by dougmoscrop** [comment]
What about using

`bucket:` as the key, and then it's either a string (the name, as this works now)
or an object: `{ name: String, create: bool }`

so you could do

`bucket: com.bucket.deploy`

This would match the existing behaviour but with a name override and should be shorthand for:

```
bucket:
    name: com.deploy.bucket
    create: true
```

but you could override create to false. which would expect the bucket to exist.

**Comment by flomotlik** [comment]
@dougmoscrop I don't think we should create the bucket. This is already a more advanced config for people who know which bucket to use in which region. Adding an SDK call to create the bucket just makes the code much more complex and I think is unnecessary for this functionality, because if users want us to create a bucket they can simply use the CF bucket name.

File: lib/plugins/aws/deploy/compile/functions/index.js (hunk 2 above)
Comment (DavidTanner) [inline lib/plugins/aws/deploy/compile/functions/index.js]: Validation of the bucket name has been moved to configureStack.  That puts the CloudFormation setup in one place.

File: lib/plugins/aws/deploy/compile/functions/index.js (hunk 1 above)
Comment (DavidTanner) [inline lib/plugins/aws/deploy/compile/functions/index.js]: I added a function to aws/lib/validate that can validate an S3 bucket name.

File: lib/plugins/aws/tests/validate.js [hunk 3]
```diff
@@ -20,50 +20,143 @@ describe('#validate()', () => {
...
+    it('should reject names that are to long', () => {
+      const bucketName = Array.from({ length: 64 }, () => 'j').join('');
+      return awsPlugin.validateS3BucketName(bucketName)
+        .then(() => {
+          throw new Error('Should not get here');
+        })
+        .catch(err => expect(err.message).to.contain('longer than 63 characters'));
     });
+
+    it('should reject names that are to short', () =>
```
Comment (eahefnawy) [inline lib/plugins/aws/tests/validate.js]: *too short

Comment (DavidTanner) [inline lib/plugins/aws/tests/validate.js]: Done.  Also fixed too long.

File: lib/plugins/aws/deploy/compile/functions/index.js [hunk 4]
```diff
@@ -9,185 +9,139 @@ class AwsCompileFunctions {
...
+    const artifactFilePath = this.serverless.service.package.individually ?
+      functionObject.artifact :
+      this.serverless.service.package.artifact;
+
+    if (!artifactFilePath) {
+      throw new Error(`No artifact path is set for function: ${functionName}`);
+    }
+
+    if (this.bucketName) {
+      newFunction.Properties.Code.S3Bucket = this.bucketName;
```
Comment (eahefnawy) [inline lib/plugins/aws/deploy/compile/functions/index.js]: @DavidTanner the PR is not working for me at all, I keep getting the following CF error:

```
  Serverless Error ---------------------------------------
 
     Template format error: Unresolved resource dependencies
     [ServerlessDeploymentBucket] in the Resources block
     of the template

```
After investigating it seems that `this.bucketName` is trying to reference `this.bucketName` in the `configureStack.js` file, which uses completely different context. So CF knows that there's a custom bucket, but compile function always tries to use the default bucket name, hence the unresolved CF dependency error.

Could you look into that? 😊 ... this is the config I used fyi:

```yml

service: custom-bucket-serverless

provider:
  name: aws
  region: us-west-2
  deploymentBucket: custom-bucket-serverless-deployment-2

functions:
  hello:
    handler: handler.hello

```

Comment (DavidTanner) [inline lib/plugins/aws/deploy/compile/functions/index.js]: I configured it like the artifactDirectoryName, so it is now attached to service.package.deploymentBucket.

Comment (DavidTanner) [inline lib/plugins/aws/deploy/compile/functions/index.js]: Also, I had forgotten to run the deployment locally after the last changes.  I ran them again now and everything looks as I was expecting it.

**Comment by DavidTanner** [comment]
@eahefnawy @flomotlik Master is merged in and everything still works.

**Comment by eahefnawy** [comment]
Thank you kind sir 😊 ... merged 🎉

**Comment by fakewaffle** [comment]
@eahefnawy, Do you anticipate this being released with 1.0?

**Comment by jaydeep17** [comment]
Is this available somewhere in the documentation?

**Comment by tommedema** [comment]
this feature seems to be undocumented

**Comment by ezeql** [comment]
Please add proper documentation. Thank you

**Comment by ethanherbertson** [comment]
@tommedema @ezeql I think it is documented here:

> You can specify your own S3 bucket which should be used to store all the deployment artifacts. The deploymentBucket config which is nested under provider lets you e.g. set the name or the serverSideEncryption method for this bucket

— https://serverless.com/framework/docs/providers/aws/guide/deploying#tips

**Pull request merged**